File: test/modbus/tcp.py
import py_nexus
import time


def main():
    try:
        host, port = "127.0.0.1", 5001

        print('[TEST] Create Modbus TCP Server')
        server = py_nexus.ModbusTCPServer()
        print('[OK]')

        coils = [True, False, True]
        inputs = [0x1234, 0x5678, 0xabcd]

        @server.CoilGetter(0x1100)
        def _():
            return coils[0]

        @server.CoilGetter(0x1101)
        def _():
            return coils[1]
        
        @server.CoilGetter(0x1102)
        def _():
            return coils[2]
        
        @server.CoilSetter(0x1100)
        def _(value):
            coils[0] = value

        @server.CoilSetter(0x1101)
        def _(value):
            coils[1] = value
        
        @server.CoilSetter(0x1102)
        def _(value):
            coils[2] = value
        
        @server.AnalogInputGetter(0x2200)
        def _():
            return inputs[0]
        
        @server.AnalogInputGetter(0x2201)
        def _():
            return inputs[1]
        
        @server.AnalogInputGetter(0x2202)
        def _():
            return inputs[2]
        
        @server.logger
        def _(ip, request, response):
            print(f'IP:{ip}, Req:{request.to_vector()}, Res:{response.to_vector()}')
        
        server.listen(host=host, port=port)
        print(f"Server is running on http://{host}:{port}/")
        time.sleep(0.1)

        print('[TEST] Create Modbus TCP Client')
        client = py_nexus.ModbusTCPClient(host=host, port=port)
        print('[OK]')

        print('[TEST] Read Coils')
        res = client.ReadCoils(0x1100, 3)
        if client.error() == py_nexus.ModbusError.NONE:
            assert res == coils
        else:
            print("Modbus error:", client.error())
        print('[OK]')

        print('[TEST] Write Coils')
        client.WriteMultipleCoils(0x1100, 3, [True, True, True])
        if client.error() == py_nexus.ModbusError.NONE:
            assert coils == [True, True, True]
        else:
            print("Modbus error:", client.error())
        print('[OK]')

        print('[TEST] Write Single Coils')
        value = client.WriteSingleCoil(0x1101, False)
        if client.error() == py_nexus.ModbusError.NONE:
            assert value == False
            assert coils == [True, False, True]
        else:
            print("Modbus error:", client.error())
        print('[OK]')

        print('[TEST] Read Input Registers')
        res = client.ReadInputRegisters(0x2200, 3)
        if client.error() == py_nexus.ModbusError.NONE:
            assert res == inputs
        else:
            print("Modbus error:", client.error())
        print('[OK]')

    except AssertionError as e:
        print(f'\033[91m--Assertion Error: {e}--\033[0m')
    
    else:
        print("\033[92m--Assertion Success--\033[0m")

    finally:
        server.stop()
        

if __name__ == '__main__':
    main()

File: test/modbus/tcp.cpp
#include "gtest/gtest.h"
#include "nexus/modbus/tcp/server.h"
#include "nexus/modbus/tcp/client.h"
#include "nexus/tools/await.h"
#include <thread>
#include <etl/keywords.h>

using namespace std::literals;

fun static convertToHexString(const std::vector<uint8_t>& data) {
    std::stringstream ss;
    for (const auto& byte : data) {
        // setw(2) ensures that each byte is represented with at least 2 characters
        // setfill('0') pads with leading zeros if necessary
        ss << std::setw(2) << std::setfill('0') << std::hex << static_cast<int>(byte);
    }
    // Reset stream flags and convert the stringstream to a string
    return ss.str();
}

TEST(Modbus, tcp) {
    val host = std::string("127.0.0.1");
    val port = 5001;

    var server = nexus::modbus::tcp::Server();
    var coils = std::vector<bool>({false, false});
    server.CoilGetter(0x0010, [&coils] { return coils[0];});
    server.CoilGetter(0x0011, [&coils] { return coils[1];});
    server.CoilSetter(0x0010, [&coils] (bool value) { coils[0] = value;});
    server.CoilSetter(0x0011, [&coils] (bool value) { coils[1] = value;});

    var analog_inputs = std::vector<uint16_t>({0xabcd, 0xef12});
    server.AnalogInputGetter(0x1010, [&analog_inputs] { return analog_inputs[0];});
    server.AnalogInputGetter(0x1011, [&analog_inputs] { return analog_inputs[1];});

    server.setLogger([] (const char* ip, nexus::byte_view request, nexus::byte_view response) {
        std::cout << "IP:" << ip << ", Req:" << convertToHexString(request) << ", Res:" << convertToHexString(response) << std::endl;
    });

    var future = std::async(std::launch::async, [&server, host, port] { return server.listen(host, port); });
    std::cout << "Server is running on http://" << host << ":" << port << std::endl;
    std::this_thread::sleep_for(1s);

    var client = nexus::modbus::tcp::Client(host, port);
    client.WriteMultipleCoils(0x0010, 2, {{false}, {true}});
    EXPECT_EQ(coils[0], false);
    EXPECT_EQ(coils[1], true);

    var [res, err] = client.ReadCoils(0x0010, 2);
    EXPECT_EQ(res[0], false);
    EXPECT_EQ(res[1], true);

    var [res2, err2] = client.ReadInputRegisters(0x1010, 2);
    EXPECT_EQ(res2[0], 0xabcd);
    EXPECT_EQ(res2[1], 0xef12);

    server.stop();
    var r = nexus::await | future;
    std::cout << (r == nexus::tcp::Server::Error::NONE ? "Server stop" : "Server fail to start in the first place") << int(r) << std::endl;
}